### PR TITLE
[DONT REVIEW] Add priority channels to bounded queue

### PR DESCRIPTION
**Pull request by black-adder**
The bounded queue should allow items to be prioritized so that less important items (spans) are dropped.

- [ ] tests

### CODE REVIEW

File: pkg/queue/bounded_queue.go [hunk 1]
```diff
@@ -60,11 +69,16 @@ func (q *BoundedQueue) StartConsumers(num int, consumer func(item interface{}))
...
 			for {
 				select {
-				case item := <-q.items:
-					atomic.AddInt32(&q.size, -1)
-					consumer(item)
 				case <-q.stopCh:
 					return
+				default:
+					for _, items := range q.items {
+						if item, ok := <-items; ok {
```
Comment (black-adder) [inline pkg/queue/bounded_queue.go]: this doesn't work as intended, it blocks forever. Back to the drawing board

File: pkg/queue/bounded_queue.go [hunk 2]
```diff
@@ -60,11 +69,18 @@ func (q *BoundedQueue) StartConsumers(num int, consumer func(item interface{}))
...
 				select {
-				case item := <-q.items:
-					atomic.AddInt32(&q.size, -1)
-					consumer(item)
 				case <-q.stopCh:
 					return
+				default:
+					for _, items := range q.items {
+						select {
+						case item := <-items:
```
Comment (black-adder) [inline pkg/queue/bounded_queue.go]: sooooo, this iterates over the priority queue and starting at the highest priority channel, checks if there's an item. If there is, it consumes it. This repeats for all channels.

it feels wrong

Comment (yurishkuro) [reply pkg/queue/bounded_queue.go]: It is wrong - hot cycle, no blocking.

Comment (black-adder) [reply pkg/queue/bounded_queue.go]: i'm gonna try using a priority queue but I feel like the all the extra syncing required will kill performance vs hot cycling.

I'll benchmark both and give you the numbers

Comment (yurishkuro) [reply pkg/queue/bounded_queue.go]: > all the extra syncing required

this is a common misconception - do you think channels are free from syncing? There are many cases where a simple mutex is a lot more efficient than achieving the same functionality with channels.

**Pull request closed without merging**